crates/hir-ty/src/object_safety.rs

```diff
@@ -0,0 +1,522 @@
+//! Compute the object-safety of a trait
+
+use std::ops::ControlFlow;
+
+use chalk_ir::{
+    cast::Cast,
+    visit::{TypeSuperVisitable, TypeVisitable, TypeVisitor},
+    BoundVar, DebruijnIndex, WhereClause,
+};
+use chalk_solve::rust_ir::InlineBound;
+use hir_def::{
+    lang_item::LangItem, AssocItemId, ConstId, FunctionId, GenericDefId, HasModule, TraitId,
+    TypeAliasId,
+};
+use rustc_hash::FxHashSet;
+
+use crate::{
+    all_super_traits,
+    db::HirDatabase,
+    from_assoc_type_id, from_chalk_trait_id,
+    generics::{generics, trait_self_param_idx},
+    lower::callable_item_sig,
+    to_assoc_type_id, to_chalk_trait_id,
+    utils::elaborate_clause_supertraits,
+    AliasEq, AliasTy, Binders, CallableSig, GoalData, ImplTraitId, Interner, OpaqueTyId,
+    ProjectionTyExt, Solution, Substitution, TraitRef, Ty, TyKind,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ObjectSafetyViolation {
+    SizedSelf,
+    SelfReferential,
+    Method(FunctionId, MethodViolationCode),
+    AssocConst(ConstId),
+    GAT(TypeAliasId),
+    // This doesn't exist in rustc, but added for better visualization
+    HasNonSafeSuperTrait(TraitId),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MethodViolationCode {
+    StaticMethod,
+    ReferencesSelfInput,
+    ReferencesSelfOutput,
+    ReferencesImplTraitInTrait,
+    AsyncFn,
+    WhereClauseReferencesSelf,
+    Generic,
+    UndispatchableReceiver,
+}
+
+// Basically, this is almost same as `rustc_trait_selection::traits::object_safety`
+// but some difference;
+//
+// 1. While rustc gathers almost every violation, but this only early return on
+//    first violation for perf.
```

Comment:
Might be nice to make this take a (generic) callback:
```rs
pub fn object_safety_of_trait_query<T>(
    db: &dyn HirDatabase,
    trait_: TraitId,
    cb: &mut dyn FnMut(ObjectSafetyViolation) -> ControlFlow<()>
) {
```
even if we don't yet need to collect all reasons

Reply:
Well, callback strategy sounds great, but I think that it can't be used as a salsa query 🤔

Reply:
Oh no it can't, mainly meant we can already architecture it that way, for the query we'd keep the single one that just calls into this with a `ControlFlow::Break` on first issue found. Anything that wants all of them can bypass the query if the query reports there is at least on violation.

Reply:
Oh, I see. I couldn't think of adding just one more "wrapper" function for query 😅